### PR TITLE
Make BackStackScreen generic

### PR DESCRIPTION
**Pull request by AquaGeek**
Part of #1046

**Pros**
* This makes `BackStackScreen` more consistent with our other containers.
* For back stacks that only have one screen type (e.g. single-screen stacks), there isn't an additional `AnyScreen` wrapping layer anymore.
* Potentially easier testing — you can get at the underlying type without having to cast.

**Cons**
* This creates more work when concatenating multiple back stacks together (which could arguably be pushed into the concatenation API (e.g. `appending<T: Screen>(_ other: BackStack<T>) -> BackStackScreen<AnyScreen>`)).
* More work is required to wrap your screen in an `AnyScreen` before adding it to a `BackStackScreen<AnyScreen>`. ~This could be mitigated by adding helper functions (e.g. in `Screen`: `func asAnyScreen() -> AnyScreen`); otherwise you have to prepend the `AnyScreen(` and reindent.~ The `asAnyScreen()` convenience has been added.

### CODE REVIEW

File: swift/Samples/TicTacToe/Sources/Game/RunGameWorkflow.swift [hunk 1]
```diff
@@ -84,19 +84,19 @@ extension RunGameWorkflow {
 // MARK: Rendering
 
 extension RunGameWorkflow {
-    typealias Rendering = AlertContainerScreen<ModalContainerScreen<BackStackScreen>>
+    typealias Rendering = AlertContainerScreen<ModalContainerScreen<BackStackScreen<AnyScreen>>>
```
Comment (dhavalshreyas) [inline swift/Samples/TicTacToe/Sources/Game/RunGameWorkflow.swift:87]: 😮

Comment (AquaGeek) [reply swift/Samples/TicTacToe/Sources/Game/RunGameWorkflow.swift:87]: Yeah…

**Comment by JustinDSN** [comment]
This is more consistent with our other open source containers where we push the wrapping of `AnyScreen` down to the client side code (eg: `SplitScreenContainer`), which at that time we thought was the right direction to take (with that container).

What are the pros & cons of doing it this way?

**Comment by dhavalshreyas** [comment]
Having thought about this more, this definitely feels like the better direction. 

Issues with being able to test `AnyScreen` will also be handled here, since we can use a different comparable wrapper instead.

File: swift/Samples/TicTacToe/Sources/Game/RunGameWorkflow.swift (hunk 1 above)
Comment (rjrjr) [reply swift/Samples/TicTacToe/Sources/Game/RunGameWorkflow.swift:87]: What's the 😮 ?

Comment (dhavalshreyas) [reply swift/Samples/TicTacToe/Sources/Game/RunGameWorkflow.swift:87]: Definitely the right direction as far as type-safety is concerned. Not ideal from the API ergonomics perspective. However, once we bump up to iOS 13 we'll be able to use [opaque types](https://docs.swift.org/swift-book/LanguageGuide/OpaqueTypes.html) and this won't be a concern.

Comment (AquaGeek) [reply swift/Samples/TicTacToe/Sources/Game/RunGameWorkflow.swift:87]: I think it's a comment on the # of generics `A<B<C<D>>>`.

File: swift/Samples/TicTacToe/Sources/Game/RunGameWorkflow.swift [hunk 2]
```diff
@@ -84,19 +84,19 @@ extension RunGameWorkflow {
...
 
     func render(state: RunGameWorkflow.State, context: RenderContext<RunGameWorkflow>) -> Rendering {
         let sink = context.makeSink(of: Action.self)
         var modals: [ModalContainerScreenModal] = []
         var alert: Alert?
 
-        var backStackItems: [BackStackScreen.Item] = [BackStackScreen.Item(
-            screen: newGameScreen(
+        var backStackItems: [BackStackScreen<AnyScreen>.Item] = [BackStackScreen.Item(
+            screen: AnyScreen(newGameScreen(
```
Comment (AquaGeek) [inline swift/Samples/TicTacToe/Sources/Game/RunGameWorkflow.swift:95]: This probably warrants a `asAnyScreen()` helper on `Screen`. The code here would then become:

```swift
screen: newGameScreen(
    sink: sink,
    playerX: state.playerX,
    playerO: state.playerO
).asAnyScreen(),
```

Comment (AquaGeek) [reply swift/Samples/TicTacToe/Sources/Game/RunGameWorkflow.swift:95]: Or I guess we could add an extension:

```swift
extension BackStackScreen where Content == AnyScreen {
    init<S: Screen>(…) {
        // Do the wrapping into AnyScreen here
    }
}
```

File: swift/Samples/Tutorial/Frameworks/Tutorial5Complete/Tests/TodoWorkflowTests.swift [hunk 3]
```diff
@@ -77,21 +78,21 @@ class TodoWorkflowTests: XCTestCase {
                     // We always expect the TodoListWorkflow
                     ExpectedWorkflow(
                         type: TodoListWorkflow.self,
-                        rendering: BackStackScreen.Item(
+                        rendering: BackStackScreen<AnyScreen>.Item(
```
Comment (JustinDSN) [inline swift/Samples/Tutorial/Frameworks/Tutorial5Complete/Tests/TodoWorkflowTests.swift:81]: As we talked about in our 1:1, I think we can snip these `<AnyScreen>` and let the compiler infer the generic type.

**Pull request merged**